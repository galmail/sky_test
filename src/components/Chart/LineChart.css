/* LineChart.css */

.line-chart {
  float: left;
  width: 760px;
  height: 300px;
  text-align: center;
  line-height: 100px;
}

.line-chart .line {
  fill: none;
  stroke: steelblue;
  stroke-width: 2px;
}

.line-chart .focus circle {
  fill: #F1F3F3;
  stroke: rgb(0, 0, 0);
  stroke-width: 5px;
}
