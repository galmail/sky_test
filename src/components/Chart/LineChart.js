import React, { Component } from 'react';
import PropTypes from 'prop-types';
import * as d3 from 'd3/build/d3';

import './LineChart.css';

class LineChart extends Component {

    constructor(props){
        super(props);
        this.rootElement = null;
        this.highlightIndicator = this.highlightIndicator.bind(this);
        this.renderChart = this.renderChart.bind(this);
    }

    render() {
        return (
            <div ref={(el) => { this.rootElement = el; }} className="line-chart">
                The Chart is Loading...
            </div>
        );
    }

    componentDidUpdate() {
        // console.log('LineChart componentDidUpdate', this.props.data);
        if (this.props.data && this.props.data.length > 0) this.renderChart(this.props.data);
        else this.rootElement.innerHTML = 'No data to display';
    }

    renderChart(data) {
        this.rootElement.innerHTML = '';
        if (!data) return;

        // set the dimensions and margins of the chart
        const margin = this.props.margin;
        const width = this.props.width - margin.left - margin.right;
        const height = this.props.height - margin.top - margin.bottom;

        const totalWidth = width + margin.left + margin.right;
        const totalHeight = height + margin.top + margin.bottom;

        // set the ranges
        const x = d3.scaleTime().range([0, width]);
        const y = d3.scaleLinear().range([height, 0]);

        // define the line
        const line = d3.line()
            .x(d => x(d.date))
            .y(d => y(d.value));

        const svg = d3.select(this.rootElement).append("svg")
            .attr("width", totalWidth)
            .attr("height", totalHeight)
            .append("g")
            .attr("transform",
            "translate(" + margin.left + "," + margin.top + ")");

        // Scale the range of the data
        x.domain(d3.extent(data, function (d) { return d.date; }));
        y.domain([0, d3.max(data, function (d) { return d.value; })]);

        // Add the valueline path.
        svg.append("path")
            .data([data])
            .attr("class", "line")
            .attr("d", line);

        // Add the X Axis
        svg.append("g")
            .attr("class", "axis-x")
            .attr("transform", "translate(0," + height + ")")
            .call(d3.axisBottom(x));

        // Add the Y Axis
        svg.append("g")
            .attr("class", "axis-y")
            .call(d3.axisLeft(y));

        // Add Circle
        const focus = svg.append("g")
            .attr("class", "focus")
            .style("display", "none");

        focus.append("circle").attr("r", 7.5);

        this.highlightIndicator(focus, x, y);
    }

    highlightIndicator(el, x, y) {
        const dataObj = this.props.highlightValue;
        const display = dataObj ? null : 'none';
        if (dataObj) el.attr("transform", "translate(" + x(dataObj.date) + "," + y(dataObj.value) + ")");
        el.style("display", display);
    }

}

LineChart.propTypes = {
    margin: PropTypes.shape({
        top: PropTypes.number,
        right: PropTypes.number,
        bottom: PropTypes.number,
        left: PropTypes.number
    }),
    width: PropTypes.number,
    height: PropTypes.number,
    data: PropTypes.arrayOf(PropTypes.shape({
        date: PropTypes.instanceOf(Date),
        value: PropTypes.number
    })),
    highlightValue: PropTypes.shape({
        date: PropTypes.instanceOf(Date),
        value: PropTypes.number
    })
};

LineChart.defaultProps = {
    margin: { top: 20, right: 20, bottom: 30, left: 50 },
    width: 760,
    height: 300
}

export default LineChart;
