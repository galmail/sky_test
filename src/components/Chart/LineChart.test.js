import React from 'react';
import { shallow, mount, render } from 'enzyme';
import moment from 'moment';

import LineChart from './LineChart';

describe('Test LineChart Component', () => {

    const DATETIME_FORMAT = 'YYYY-MM-DDThh:mm';
    const data = [
        { date: moment('2017-12-10T10:00', DATETIME_FORMAT).toDate(), value: 30.0 },
        { date: moment('2017-12-10T11:00', DATETIME_FORMAT).toDate(), value: 25.0 },
        { date: moment('2017-12-10T12:00', DATETIME_FORMAT).toDate(), value: 10.0 },
    ];

    it('should render without crashing', () => {
        const wrapper = shallow(<LineChart />);
        expect(wrapper).toMatchSnapshot();
    });

    it('is calling the right methods when updating', (done) => {
        const wrapper = mount(<LineChart />);
        let spy = spyOn(wrapper.instance(), 'highlightIndicator');
        wrapper.setProps({ data: data, highlightValue: data[0] }, () => {
            expect(spy).toHaveBeenCalled();
            spy = spyOn(wrapper.instance(), 'renderChart');
            wrapper.setProps({ data: data, highlightValue: data[1] }, () => {
                expect(spy).toHaveBeenCalled();
                expect(spy).toHaveBeenCalledWith(data);
                done();
            });
        });
    });

    it('should be able to highlight data', () => {
        const wrapper = mount(<LineChart data={data} highlightValue={data[2]} />);
        wrapper.instance().renderChart(data);
        expect(wrapper.render().find('.focus').length).toBe(1);
        expect(wrapper.render().find('.focus').prop('style').display).toBe(undefined);
        expect(wrapper.render().find('.focus').prop('transform')).toBe('translate(690,166.66666666666669)');
    });

});
