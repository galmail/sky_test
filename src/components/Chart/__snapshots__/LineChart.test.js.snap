exports[`Test LineChart Component should render without crashing 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <LineChart
    height={300}
    margin={
        Object {
            "bottom": 30,
            "left": 50,
            "right": 20,
            "top": 20,
          }
    }
    width={760} />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": "The Chart is Loading...",
      "className": "line-chart",
    },
    "ref": [Function],
    "rendered": "The Chart is Loading...",
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": "The Chart is Loading...",
        "className": "line-chart",
      },
      "ref": [Function],
      "rendered": "The Chart is Loading...",
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactFifteenAdapter {
      "options": Object {
        "supportPrevContextArgumentOfComponentDidUpdate": true,
      },
    },
  },
}
`;
